### PR TITLE
Allow missions to mark systems of importance

### PR DESCRIPTION
**Pull request by Amazinite**
**Feature**

This PR addresses the feature described in issue #5145

## Summary
Adds a new keyword for missions:
* `"mark" <system>`: The system with the given name will be marked when this mission is active. Marked systems don't need to be visited in order for the mission to be completed (unlike waypoints), but may be points of interest for the mission, such as a location of a side objective.

Currently, marks display the same as stopovers or waypoints. (There's not enough space in the mission marker key to add another line for a new color.) As such, I've renamed the last line in the mission key from "Waypoint you must visit" to "System of importance"; how that system is important to the mission should be made clear by the mission description.

Marks can be referred to in mission text with a `<marks>` replacement similar to `<waypoints>`.

Marks are not removed when the system is visited. They can only be removed by the "unmark" keyword in a GameAction.
* `"unmark" <system>`: Causes the mark to be removed from the map. Selecting the mission in on the map will still draw circles around the unmarked systems like it would visited waypoints if for some reason the player still needs to review which systems where a part of the mission.

New marks can also be added by actions with the "mark" keyword:
* `"mark" <system>`: Causes the system to be marked.

## Screenshots
It looks just like a waypoint ping, but behaves differently.
image
I've entered the system, but the mark is still present.
image
Landing on Earth completes the mission regardless of whether the mark was visited.
image
If I unmark the system in an action, the system is still visible as important to the mission by selecting the mission on the map.
image
The changed map key. (Updated the text for the last line.)
image

## Usage examples
The following mission would mark the Alpha Centauri system as a location of interest for the mission. You could go to the destination and complete the mission without needing to go to Alpha Centauri. Upon entering Alpha Centauri, the "on enter" action unmarks the system. If this action were not present, the system would stay marked even after entering it.
```
mission "Mark Test"
	job
	description "Travel to Earth. Check out Alpha Centauri before you land if you want."
	mark "Alpha Centauri"
	destination "Earth"
	on enter "Alpha Centauri"
		unmark "Alpha Centauri"
		mark "Wei"
		dialog "You enter the system. There's not actually anything important here. There might be something over at Wei, though."
	on enter "Wei"
		unmark "Wei"
		dialog "There was nothing here either."
```

## Testing Done
Tested the above mission. Observed that I could finish the mission without needing to go to Alpha Centauri, and that entering the system didn't unmark it if there are no actions to do so. Upon adding an action to do so, Alpha Centauri became unmarked when entering the system while Wei became marked. Flying to Wei unmarked it.

### CODE REVIEW

**Comment by Hurleveur** [comment]
I love the idea and already know where to make use of it, but that said it should probably be made different from a waypoint in some form or another, from a player perspective, else it might be a bit confusing.
If we cant change the color we should probably change something else, say by making it transparent or smth

**Comment by bene-dictator** [comment]
> I love the idea and already know where to make use of it, but that said it should probably be made different from a waypoint in some form or another, from a player perspective, else it might be a bit confusing.
If we cant change the color we should probably change something else, say by making it transparent or smth

Not always, because that may be the intention. Say we want to make the Pact Recon 3 mission to only check when you've gone to Naper; we want to make the other arrows look the same as to not give away the fact that the pirate fleet they're looking for is in Naper.

If the idea is considered, at least keep the original arrow.

**Comment by Hurleveur** [comment]
The issue suggests to make the marker duller after having seen it once, at least that should be done, I think

**Comment by demolish238** [comment]
This is something I've had to make workarounds for with two of my plugins, will be great to just be able to ping points of interest properly.

**Comment by warp-core** [comment]
I suspect the reason unpinging isn't working is that you're not copying the set of systems to unping into the new GameAction in GameAction::Instantiate.

**Comment by Amazinite** [comment]
Many such cases. ~~I could've sworn I checked there.~~

File: source/MapPanel.cpp [hunk 1]
```diff
@@ -1471,6 +1479,8 @@ void MapPanel::DrawMissions()
 			DrawPointer(waypoint, missionCount[waypoint].drawn, waypointColor);
 		for(const Planet *stopover : mission.Stopovers())
 			DrawPointer(stopover->GetSystem(), missionCount[stopover->GetSystem()].drawn, waypointColor);
+		for(const System *ping : mission.Pings())
+			DrawPointer(ping, missionCount[ping].drawn, waypointColor);
```
Comment (TomGoodIdea) [inline source/MapPanel.cpp:1483]: When you both set a system as a waypoint and ping it, you get two dark blue markers from the same mission on a single system. It doesn't make much sense to me, unless we're going to make pings and waypoints look different from each other.

**Review by TomGoodIdea — COMMENTED** [review]
I think we should avoid having too much different markers, because as demolish238 said, it may become confusing. The ring around an unpinged system could be more visible, though.

**Comment by warp-core** [comment]
I also think this should be called "mark" instead of "ping."
As in "mark system". Since that's what it's doing.
"Ping" seems more like it should relate to a transient event of some sort. Not something that persists until it is turned off.

File: source/MapPanel.cpp (hunk 1 above)
Comment (Amazinite) [reply source/MapPanel.cpp:1483]: I'd question why a mission would both mark/ping a system and use it as a waypoint in the first place. Even if there would be a good reason to waypoint and mark a system, I believe a system that is a waypoint and contains a stopover will also show two markers of the same type; should that also change, or how would that case be different where it shouldn't change?

Comment (TomGoodIdea) [reply source/MapPanel.cpp:1483]: Hmm, that's a good point; it can be changed in a separate PR if we want.

**Review by ziproot — APPROVED** [review]

**Pull request merged**